pulsar-broker/src/main/java/com/yahoo/pulsar/broker/loadbalance/impl/ModularLoadManagerImpl.java

```diff
@@ -414,18 +414,28 @@ private void updateBundleData() {
 
             // Remove all loaded bundles from the preallocated maps.
             final Map<String, BundleData> preallocatedBundleData = brokerData.getPreallocatedBundleData();
-            // Should not iterate with more than one thread at a time.
             synchronized (preallocatedBundleData) {
-                final Iterator<Map.Entry<String, BundleData>> preallocatedIterator = preallocatedBundleData.entrySet()
-                        .iterator();
-                while (preallocatedIterator.hasNext()) {
-                    final String bundle = preallocatedIterator.next().getKey();
-                    if (bundleData.containsKey(bundle)) {
-                        preallocatedIterator.remove();
-                        preallocatedBundleToBroker.remove(bundle);
+                for (String bundleName : brokerData.getLocalData().getBundles()) {
+                    if (preallocatedBundleData.containsKey(bundleName)) {
+                        if (log.isDebugEnabled()) {
+                            log.debug("Preallocated bundle {[]} is now loaded on broker [{}]", bundleName, broker);
+                        }
+                        final Iterator<Map.Entry<String, BundleData>> preallocatedIterator = preallocatedBundleData.entrySet()
+                                .iterator();
+                        while (preallocatedIterator.hasNext()) {
```

Comment:
Here, I think we are iterating on `preallocatedBundleData` for every single bundle of `brokerData`. Instead should we keep earlier code same and before updating `preallocatedBundleData`, we can remove all the bundles which are not part of `brokerData`. So, code will be simple to understand and both collections are map so, it can be done in O(n) iteration only. any thought?